### ISSUE TITLE
api to find an available port

### ISSUE THREAD

**Issue by timotheecour** [vibe-d/vibe.d#1431]
would be nice to have an api to find an available port so user can use that instead of hard coding a port
there's some non-trivial cross platform logic to figure out (eg valid port range depends on platform + other settings eg: http://stackoverflow.com/questions/913501/how-to-let-kernel-choose-a-port-number-in-the-range-1024-5000-in-tcp-socket-pr)

(alternatively, but less preferably, let 0 mean find an available port as in http://stackoverflow.com/questions/2675362/how-to-find-an-available-port for ServerSocket(0), but would still be nice to allow specifying an optional range in which to find a port)

on a related note, looks like starting a vibe server on an already used port doesn't complain, why is that?

**Comment by s-ludwig** [comment]
Have you tried passing port 0? I've never tried it, but I would have expected it to just work. Since an API for a specific range would be rather inefficient and since it can be achieved by calling the existing API repeatedly, I'm a little reluctant to add one.

**Comment by timotheecour** [comment]
I can specify port=0 but how would i get the port actually assigned after passing port 0? There should be an API for that as well (eg so user can browse to $host:$port)
Also, the disadvantage is it's less flexible (can't pass an application specific range)

>> rather inefficient and since it can be achieved by calling the existing API repeatedly

which API?
Calling listenHTTP repeatedly incurs significant overhead as it does a whole bunch of other things.

**Comment by s-ludwig** [comment]
You can use `TCPConnection.localAddress.port` to get the bind port.

AFAIK, there is no OS API for specifying a specific range of ports, so calling bind repeatedly is the only option anyway. I agree that calling `listenTCP` incurs even more overhead, because it also creates a new socket each time, but it wouldn't be efficient anyway.

**Comment by timotheecour** [comment]
>> You can use TCPConnection.localAddress.port to get the bind port.

do you have an example?
```
auto settings = new HTTPServerSettings;
settings.port = 0;
settings.bindAddresses = ["::1", "127.0.0.1"];
auto listener=listenHTTP(settings, &handler);

// listener only has a single public method stopListening; 
// how to get bind port?
```

>> AFAIK, there is no OS API for specifying a specific range of ports

OK, then I will create my own based on system calls.

**Comment by s-ludwig** [comment]
>>  You can use TCPConnection.localAddress.port to get the bind port.

> do you have an example?

Sorry, I got confused there. I've added `TCPListener.bindAddress` now.

**Comment by s-ludwig** [comment]
I'd consider this resolved by #1431, as querying a free port separately from the actual `bind` call is an unnecessary source for race-conditions.
